instrumentation/github.com/go-kit/kit/otelkit/endpoint.go

```diff
@@ -0,0 +1,135 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Based on https://github.com/go-kit/kit/blob/3796a6b25f5c6c545454d3ed7187c4ced258083d/tracing/opencensus/endpoint.go
+
+package otelkit
+
+import (
+	"context"
+	"strconv"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/label"
+	"go.opentelemetry.io/otel/trace"
+
+	otelcontrib "go.opentelemetry.io/contrib"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/sd/lb"
+)
+
+const (
+	tracerName = "go.opentelemetry.io/contrib/instrumentation/github.com/go-kit/kit/otelkit"
+
+	// defaultSpanName is the default endpoint span name to use.
+	defaultSpanName = "gokit/endpoint"
+)
+
+// EndpointMiddleware returns an Endpoint middleware, tracing a Go kit endpoint.
+// This endpoint tracer should be used in combination with a Go kit Transport
+// tracing middleware, generic OpenCensus transport middleware or custom before
+// and after transport functions as service propagation of SpanContext is not
+// provided in this middleware.
+func EndpointMiddleware(options ...EndpointOption) endpoint.Middleware {
+	cfg := &endpointOptions{}
+
+	for _, o := range options {
+		o(cfg)
+	}
+
+	if cfg.TracerProvider == nil {
+		cfg.TracerProvider = otel.GetTracerProvider()
+	}
+
+	tracer := cfg.TracerProvider.Tracer(
+		tracerName,
+		trace.WithInstrumentationVersion(otelcontrib.SemVersion()),
+	)
+
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+			operation := cfg.Operation
+			if cfg.GetOperation != nil {
+				if newOperation := cfg.GetOperation(ctx, operation); newOperation != "" {
+					operation = newOperation
+				}
+			}
+
+			spanName := operation
+			if spanName == "" {
+				spanName = defaultSpanName
+			}
+
+			opts := []trace.SpanOption{
+				trace.WithAttributes(cfg.Attributes...),
+				trace.WithSpanKind(trace.SpanKindServer),
+			}
+
+			if cfg.GetAttributes != nil {
+				opts = append(opts, trace.WithAttributes(cfg.GetAttributes(ctx)...))
+			}
+
+			ctx, span := tracer.Start(ctx, spanName, opts...)
+			defer span.End()
+
+			defer func() {
+				if err != nil {
+					if lberr, ok := err.(lb.RetryError); ok {
+						// handle errors originating from lb.Retry
+						for idx, rawErr := range lberr.RawErrors {
+							span.AddEvent(
+								"error",
+								trace.WithAttributes(
+									label.String("error.type", "gokit.lb.retry"),
+									label.String("gokit.lb.retry.count", strconv.Itoa(idx+1)),
```

Comment:
Could this be `label.Int` instead or is there a reason for converting to string?